syft/src/main/java/org/openmined/syft/data/DataLoader.kt

```diff
@@ -23,43 +22,62 @@ class DataLoader(var dataset: Dataset,
                  var sampler: Sampler = SequentialSampler(dataset),
                  var batchSize: Int = 1,
                  var dropLast: Boolean = false
-) : Iterable<Pair<IValue, IValue>> {
+) : Iterable<List<IValue>> {
 
-    private val batchSampler = BatchSampler(
+    val indexSampler = BatchSampler(
         sampler,
         batchSize,
         dropLast
     )
 
-    fun indexSampler() = batchSampler
-
     private val iterator = DataLoaderIterator(this)
 
-    override fun iterator(): Iterator<Pair<IValue, IValue>> = iterator
+    override fun iterator(): Iterator<List<IValue>> = iterator
 
     fun reset() {
-        indexSampler().reset()
+        indexSampler.reset()
         iterator.reset()
     }
 }
 
-class DataLoaderIterator(dataLoader: DataLoader) : Iterator<Pair<IValue, IValue>> {
-
-    private val indexSampler = dataLoader.indexSampler()
+class DataLoaderIterator(dataLoader: DataLoader) : Iterator<List<IValue>> {
 
-    private val datasetFetcher = BaseDatasetFetcher(dataLoader.dataset)
+    private val indexSampler = dataLoader.indexSampler
 
-    private val datasetLength = datasetFetcher.dataset.length()
+    private val dataset = dataLoader.dataset
 
     private var currentIndex = 0
 
-    override fun next(): Pair<IValue, IValue> {
-        val indices = indexSampler.indices()
+    override fun next(): List<IValue> {
+        val indices = indexSampler.indices
         currentIndex += indices.size
-        return datasetFetcher.fetch(indices)
+        return fetch(indices)
+    }
+
+    private fun fetch(indices: List<Int>): List<IValue> {
```

Comment:
Fetch function should be a part of dataloader and should be overridable by the user